runtime/runc_jailer.go

```diff
@@ -0,0 +1,351 @@
+			}
+
+			rootPathToConfig := filepath.Join(ociBundlePath, "config.json")
+			j.logger.WithField("rootPathToConfig", rootPathToConfig).Debug("Copying config")
+			if err := copyFile(runcConfigPath, rootPathToConfig, 0444); err != nil {
+				return errors.Wrapf(err, "failed to copy config from %v to %v", runcConfigPath, rootPathToConfig)
+			}
+
+			j.logger.Debug("Overwritting process args of config")
+			if err := j.overwriteConfig(cfg, filepath.Base(m.Cfg.SocketPath), rootPathToConfig); err != nil {
+				return errors.Wrap(err, "failed to overwrite config.json")
+			}
+
+			// copy the firecracker binary
+			j.logger.WithField("root path", rootPath).Debug("copying firecracker binary")
+			newFirecrackerBinPath := filepath.Join(rootPath, filepath.Base(cfg.FirecrackerBinaryPath))
+			if err := copyFile(
+				cfg.FirecrackerBinaryPath,
+				newFirecrackerBinPath,
+				0700,
+			); err != nil {
+				return errors.Wrapf(err, "could not copy firecracker binary from path %v", cfg.FirecrackerBinaryPath)
+			}
+			if err := os.Chown(newFirecrackerBinPath, int(j.uid), int(j.gid)); err != nil {
+				return errors.Wrap(err, "failed to change ownership of binary")
+			}
+
+			// Link the kernel image
+			newKernelImagePath := filepath.Join(rootPath, kernelImageFileName)
+			j.logger.WithField("newKernelImagePath", newKernelImagePath).Debug("linking kernel image")
+			if err := os.Link(m.Cfg.KernelImagePath, newKernelImagePath); err != nil {
+				return errors.Wrapf(err, "failed to link kernel image from path %v", m.Cfg.KernelImagePath)
+			}
+			m.Cfg.KernelImagePath = kernelImageFileName
+
+			// link drives to new contents path
+			for i, d := range m.Cfg.Drives {
+				drivePath := firecracker.StringValue(d.PathOnHost)
+				fileName := filepath.Base(drivePath)
+				newDrivePath := filepath.Join(rootPath, fileName)
+
+				if err := os.Link(drivePath, newDrivePath); err != nil {
+					return errors.Wrapf(err, "failed to copy drive %v", drivePath)
+				}
+
+				f, err := os.Open(drivePath)
+				if err != nil {
+					return errors.Wrap(err, "failed to open drive file")
+				}
+
+				if internal.IsStubDrive(f) {
+					j.logger.Debug("stub drive detected")
+					if err := os.Chown(newDrivePath, int(j.uid), int(j.gid)); err != nil {
+						if e := f.Close(); e != nil {
+							j.logger.WithError(e).Debug("failed to close drive file")
+						}
+						return errors.Wrap(err, "failed to change ownership of stub drive")
+					}
+				}
+
+				if err := f.Close(); err != nil {
+					j.logger.WithError(err).Debug("failed to close drive file")
+				}
+
+				j.logger.WithField("drive", newDrivePath).Debug("Adding drive")
+				m.Cfg.Drives[i].PathOnHost = firecracker.String(fileName)
+			}
+
+			// Setting the proper path to where the vsock path should be
+			for i, v := range m.Cfg.VsockDevices {
+				filename := filepath.Base(v.Path)
+				v.Path = filepath.Join("/", filename)
+				m.Cfg.VsockDevices[i] = v
```

Comment:
Would probably be nice to have some debug logging indicating what's going on here, too.

Reply:
Resolved?